Backpack/BottomSheet/Classes/BottomSheet.swift

```diff
@@ -148,50 +162,70 @@ public final class BPKBottomSheet: NSObject {
         bottomSheet.present(in: floatingPanelController, animated: animated, completion: completion)
     }
 
+    /// This method should be used if you inted to show the BPKBottomSheet and keep tracking the backdrop view
+    /// user interactions. If you need your BPKBottomSheet have a modal behaviour you should use present method
+    /// should be used instead.
+    /// - Parameter parent: Parent view where the BPKBottomSheet will be placed.
+    public func addPanel(toParent parent: UIViewController) {
+        assert(presentationStyle == .persistent, "present(in:animated:completion:) not compatible with non modal presentation style")
```

Comment:
I'm not stoked about these asserts, maybe we can utilise generic and optional conformance to have the type system solve this for us intead of using runtime errors with asserts? 

i.e. something like:

```swift
protocol BPKBottomSheetPresentationStyle {}
protocol BPKBottomSheetPresentationStyleModal: BPKBottomSheetPresentationStyle {}
protocol BPKBottomSheetPresentationStylePersistent: BPKBottomSheetPresentationStyle {}

 public final class BPKBottomSheet<Style: BPKBottomSheetPresentationStyle> : NSObject { 
....
}

extension BPKBottomSheet where Style: BPKBottomSheetPresentationStylePersistent {
     public func addPanel(toParent parent: UIViewController) { /* ... */ }
}

public struct _BPKBottomSheetPersistentStyle: BPKBottomSheetPresentationStyle, BPKBottomSheetPresentationStylePersistent {}

typealias BPKBottomSheetPersistent = BPKBottomSheet<_BPKBottomSheetPersistentStyle>
```

Reply:
I think if we follow this approach we would cause an API breaking change. When I opt-in for the `assert` approach, I took into consideration API breaking changes. 

If this is a hard request, I'll be happy to apply it. I don't have any other ideas right now about implement it in a different fashion where I would avoid asserting types. Your idea seems to work here and I would follow a similar path as you suggested.

Reply:
> I think if we follow this approach we would cause an API breaking change. When I opt-in for the assert approach, I took into consideration API breaking changes.

Yeah that's a good instinct, but often it's okay to make a breaking change if it improves the API greatly(imo). Here we need to balance the worse feedback loop with asserts vs a the improved feedback loop when using an assert instead. 

I think we probably shouldn't do this and live with the asserts. One thing to note though is that in a production build the asserts will not do anything and we will still take the action. Should we ensure that in production builds we just do nothing?

Reply:
We might get this behavior using `assertionFailure`-> 8a52851727adef325214ff87cb0bcb8a8ed64e1e

It will fail in Debug builds but will do nothing in prod builds.

By documentation:

```Swift
/// Indicates that an internal sanity check failed.
///
/// Use this function to stop the program, without impacting the performance of
/// shipping code, when control flow is not expected to reach the call---for
/// example, in the `default` case of a `switch` where you have knowledge that
/// one of the other cases must be satisfied. To protect code from invalid
/// usage in Release builds, see `preconditionFailure(_:file:line:)`.
///
/// * In playgrounds and -Onone builds (the default for Xcode's Debug
///   configuration), stop program execution in a debuggable state after
///   printing `message`.
///
/// * In -O builds, has no effect.
///
/// * In -Ounchecked builds, the optimizer may assume that this function is
///   never called. Failure to satisfy that assumption is a serious
///   programming error.
///
/// - Parameters:
///   - message: A string to print in a playground or `-Onone` build. The
///     default is an empty string.
///   - file: The file name to print with `message`. The default is the file
///     where `assertionFailure(_:file:line:)` is called.
///   - line: The line number to print along with `message`. The default is the
///     line number where `assertionFailure(_:file:line:)` is called.
```

Reply:
Yeah the way you have rewritten it now works 👍🏼